### PR TITLE
remove recursive calls

### PR DESCRIPTION
**Pull request by segfaultdoc**
Fixes the issue mentioned [here](https://github.com/nats-io/nats.rs/issues/446#issue-1242024070)

### CODE REVIEW

File: nats/src/jetstream/push_subscription.rs [hunk 1]
```diff
@@ -188,22 +192,24 @@ impl PushSubscription {
     /// # }
     /// ```
     pub fn next_timeout(&self, timeout: Duration) -> io::Result<Message> {
-        match self.0.messages.recv_timeout(timeout) {
-            Ok(message) => {
-                if self.preprocess(&message) {
-                    return self.next_timeout(timeout);
-                }
+        loop {
+            return match self.0.messages.recv_timeout(timeout) {
```
Comment (buffalu) [inline nats/src/jetstream/push_subscription.rs:196]: same here?

File: nats/src/jetstream/pull_subscription.rs [hunk 2]
```diff
@@ -316,27 +316,29 @@ impl PullSubscription {
     /// # }
     /// ```
     pub fn next_timeout(&self, timeout: Duration) -> io::Result<Message> {
-        match self.0.messages.recv_timeout(timeout) {
-            Ok(message) => {
-                if message.is_no_messages() {
-                    return self.next_timeout(timeout);
+        loop {
```
Comment (whistlinwilly) [inline nats/src/jetstream/pull_subscription.rs:319]: smart using loop over recursion here

File: nats/src/jetstream/pull_subscription.rs [hunk 3]
```diff
@@ -316,27 +316,29 @@ impl PullSubscription {
...
+            return match self.0.messages.recv_timeout(timeout) {
+                Ok(message) => {
+                    if message.is_no_messages() {
```
[the remark this answers is not in the document]
Comment (whistlinwilly) [reply nats/src/jetstream/pull_subscription.rs:322]: timeout would get reset on the recursive call i think?

**Review by whistlinwilly — APPROVED** [review]
fuck yes nice find and fix

File: nats/src/jetstream/pull_subscription.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (segfaultdoc) [reply nats/src/jetstream/pull_subscription.rs:322]: hm yea I didn't want to change the functionality of the code, this works the same way as recursive impl, you think that should be implemented?

File: nats/src/kv.rs [hunk 4]
```diff
@@ -784,29 +784,31 @@ impl Iterator for Keys {
...
-                    if info.pending == 0 {
-                        self.done = true;
+        loop {
+            return match self.subscription.next() {
+                Some(message) => {
+                    // If there are no more pending messages we'll stop after delivering the key
+                    // derived from this message.
+                    if let Some(info) = message.jetstream_message_info() {
+                        if info.pending == 0 {
+                            self.done = true;
```
Comment (segfaultdoc) [inline nats/src/kv.rs:794]: return here

**Review by caspervonb — COMMENTED** [review]
Nice find and looks good! Will wait for Thomasz to have a look before merging.

File: nats/src/jetstream/pull_subscription.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jarema) [reply nats/src/jetstream/pull_subscription.rs:322]: Good catch!

Resetting the timeout was definitely unintended and is abug. User should not have to care if there are some conditions inside the function that can reset the timeout. If the timeout is reached, the function should return.

[the remark this answers is not in the document]
Comment (segfaultdoc) [reply nats/src/jetstream/pull_subscription.rs:322]: updated

**Review by Jarema — APPROVED** [review]
Thanks a lot for the contribution!

LGTM.

**Pull request merged**